text/0000-bind-helper.md

```diff
@@ -0,0 +1,124 @@
+- Start Date: 2019-03-20
+- Relevant Team(s): Ember.js, Learning
+- RFC PR: https://github.com/emberjs/rfcs/pull/470
+- Tracking: (leave this empty)
+
+# Bind Helper
+
+## Summary
+
+This RFC introduces a new helper `bind` to allow clear argument and context scope binding for functions passed in templates.
+
+## Motivation
+
+The current `action` helper has a lot of implicit and confusing behavior that is different than the Octane and post Octane programming model.
+
+To understand the complexity of `action` there are many complex behaviors including:
+
+1. Argument partial application (currying)
+2. `this` context binding
+3. `send` checks for Component and Controllers
+
+At build time the `action` helper currently is passed through an AST transform to explicitly bind the `this` to be deterministic at runtime. This is a private API where the outputted Glimmer is not a 1-1 to the template. Also, the `action` helper is confused and has overlap with the `action` modifier which has similar but slightly different behavior.
+
+Instead of this confusing and overloaded behavior, a new `bind` helper would be introduced to do partial application and context binding (with no need for build time private APIs).
+
+## Detailed design
+
+The `bind` helper will take in a function and then the set of arguments that will be partially applied to the function.
+As an optional named argument the user can pass in a `this` property that will set the `this` context this named argument would default to `undefined`.
+
+The behavior of this helper will be simple and use JavaScript's `function.prototype.bind`.
+
+Here are some examples of the `bind` helper and the equivalent JS:
+
+### Simple Case On Argument Curry
+
+```hbs
+{{bind this.log 1}}
+```
+
+```js
+this.log.bind(null, 1);
+```
+
+### Multiple Argument Partial Application
+
+```hbs
+{{bind this.add 1 2}}
+```
+
+```js
+this.add.bind(null, 1, 2);
+```
+
+### Setting the `this` context
+
+```hbs
+{{{bind this.session.logout context=this.anotherSession}}}
```

Comment:
I think `context` is a pretty overloaded word that doesn't really convey a lot of meaning. Why don't we use `thisArg` to make it less ambiguous and match the [EMCA spec](https://www.ecma-international.org/ecma-262/5.1/#sec-15.3.4.5)'s naming?

Reply:
+1, or just `this`?